main.go

```diff
@@ -0,0 +1,334 @@
+package main
+
+import (
+    "bytes"
+
+    "fmt"
+
+    "path/filepath"
+
+    "os"
+    "os/exec"
+    "os/user"
+
+    "io/ioutil"
+
+    "github.com/drone/drone-plugin-go/plugin"
+)
+
+var (
+    buildCommit string
+)
+
+var netrcFile = `
+machine %s
+login %s
+password %s
+`
+
+type Params struct {
+    UpstreamName    string            `json:"upstream"`
+    PagesDirectory  string            `json:"pagesDirectory"`
+    TemporaryBase   string            `json:"temporaryBaseDirectory"`
+    TargetBranch    string            `json:"target"`
+}
+
+func main() {
+    fmt.Printf("Drone gh-pages deployment plugin built from %s\n", buildCommit)
+
+    v := new(Params)
+    r := new(plugin.Repo)
+    b := new(plugin.Build)
+    w := new(plugin.Workspace)
+    plugin.Param("repo", r)
+    plugin.Param("build", b)
+    plugin.Param("workspace", w)
+    plugin.Param("vargs", &v)
+    plugin.MustParse()
+    err := publishDocs(r, b, w, v)
+    if err != nil {
+        fmt.Printf("%s\n", err)
+        os.Exit(1)
+    }
+}
+
+func publishDocs(r *plugin.Repo, b *plugin.Build, w *plugin.Workspace, v *Params) error {
+    if (v.UpstreamName == "") {
+        v.UpstreamName = "origin"
+    }
+    if (v.TargetBranch == "") {
+        v.TargetBranch = "gh-pages"
+    }
+    if (v.TemporaryBase == "") {
+        v.TemporaryBase = ".tmp"
+    }
+    if (v.PagesDirectory == "") {
+        v.PagesDirectory = "docs"
+    }
+
+    temporaryBaseDirectory := ""
+    if filepath.IsAbs(v.TemporaryBase) {
+        temporaryBaseDirectory = v.TemporaryBase
+    } else {
+        temporaryBaseDirectory = filepath.Join(w.Path, v.TemporaryBase)        
+    }
+    temporaryPagesDirectory := filepath.Join(temporaryBaseDirectory, v.PagesDirectory)
+
+    fullPagesDirectory := filepath.Join(w.Path, v.PagesDirectory)
+
+    // generate the .netrc file
+    if err := writeNetrc(w); err != nil {
+        fmt.Fprintln(os.Stderr, err)
+        return err
+    }
+
+    // write the rsa private key if provided
+    if err := writeKey(w); err != nil {
+        fmt.Fprintln(os.Stderr, err)
+        return err
+    }
+
+    err := runCommand(GlobalUser(b), nil)
+    if err != nil {
+        return err
+    }
+
+    err = runCommand(GlobalName(b), nil)
+    if err != nil {
+        return err
+    }
+
+    err = os.MkdirAll(temporaryPagesDirectory, 0777)
+    if err != nil {
+        return err
+    }
+
```

Comment:
you can do `defer os.RemoveAll(...)` so that you don't have to invoke twice below